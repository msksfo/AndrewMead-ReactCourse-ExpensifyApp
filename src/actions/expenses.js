import uuid from 'uuid';

//set up the action generator for adding an expense
// destructure the values that come from the user with default values in case        none are passed in
const addExpense = ({
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
} = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt,
    },
});

// set up the action generator for removing an expense by id
// also could pass in the id as a destructured value { id }
const removeExpense = id => ({
    type: 'REMOVE_EXPENSE',
    id,
});

// set up the action generator for editing an expense
// pass in an id, and an updates object
const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates,
});

export { addExpense, removeExpense, editExpense };
