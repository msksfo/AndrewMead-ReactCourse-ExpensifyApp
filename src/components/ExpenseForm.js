import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

//console.log(now.format('MMM Do, YYYY'));

// this contains all the form fields and logic to get the form to work
export default class ExpenseForm extends React.Component {
    // use local component state to track the changes to these form inputs. only when the user submits the form will we send it off to redux. redux will figure out what to do (add a new expense, or edit the existing expense)

    //! WHY DO WE USE A CONSTRUCTOR FUNCTION HERE?
    //! we need to look at the props (for our ternary conditions). so we need to use a constructor function, and define the state there. then we can pass the props up to super
    constructor(props) {
        super(props);

        // *we only want to use defaults for the first 4 values if they are CREATING an expense. if they are EDITING an expense, we need to know what their value were so we can populate the form fields with that information.

        this.state = {
            description: props.expense ? props.expense.description : '',
            noteText: props.expense ? props.expense.noteText : '',
            amount: props.expense
                ? (props.expense.amount / 100).toString()
                : '',
            createdAt: props.expense
                ? moment(props.expense.createdAt)
                : moment(),
            calendarFocused: false,
            error: '',
        };
    }

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onAmountChange = e => {
        const amount = e.target.value;
        // ensure that the number comes in a specific format, with regex

        // myRegex makes sure the value starts with a number ^\d
        // force usage of at least one number ^\d{1,}
        // optionally allow for one (only one!) decimal point  ^\d*(\.)?
        // allow only 2 numbers after the decimal point ^\d*(\.\d{0,2})?
        // make sure the regex ends with 2 numbers after the decimal point (if /when present) ^\d*(\.\d{0,2})?$

        const myRegex = /^\d{1,}(\.\d{0,2})?$/;

        // if there is no amount, OR the amount provided is a match
        if (!amount || amount.match(myRegex)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = createdAt => {
        // call this function if a date was picked
        // if it was cleared, it will be called with nothing
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    //?  I dont really understand this function. Review Andrew's explanation in section 11, lecture 106 or read docs
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onNoteChange = e => {
        //! trying to access e.target.value inside the callback to setState will not work!

        //* always make a variable for this (e.target.value)
        // the other option would be to use e.persist()
        const noteText = e.target.value;
        this.setState(() => ({ noteText }));
    };

    onSubmit = e => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            // set error state equal to 'please provide amount and description'
            this.setState(() => ({
                error: 'please provide amount and description',
            }));
        } else {
            // clear the error
            this.setState(() => ({ error: '' }));

            // submit the form
            this.props.onSubmit({
                description: this.state.description,
                // turn the string into a decimal number, and multiply by 100 to get dollars from our cents
                amount: parseFloat(this.state.amount, 10) * 100,
                // parse this to get the timestamp
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        // adding value makes it a read only input. to make it dynamic as the user types, add the onChange handler
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        placeholder="Description"
                        autoFocus
                    />

                    <input
                        type="text"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        placeholder="Amount"
                    />

                    <SingleDatePicker
                        //*the first 4 properties are mandatory
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        // make all days available, even dates in the past
                        isOutsideRange={() => false}
                    />

                    <textarea
                        placeholder="Add a note for your stupid fucking expense (fucking optional)"
                        value={this.state.noteText}
                        onChange={this.onNoteChange}
                    />

                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}
