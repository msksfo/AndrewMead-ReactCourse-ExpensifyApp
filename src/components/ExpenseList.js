import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//*  use the connect function in the individual components where you need to dispatch actions or read from the store. this connects your component to the redux store

//* when you connect a component to the redux store, it's reactive. which means that as the store changes, the component gets rerendered with the new values

// this is being exported so that we can test it
export const ExpenseList = props => (
    <div>
        {props.expenses.length === 0 ? (
            <p>no expenses</p>
        ) : (
            props.expenses.map(value => (
                <ExpenseListItem
                    key={value.id}
                    id={value.id}
                    description={value.description}
                    amount={value.amount}
                    createdAt={value.createdAt}
                />
            ))
        )}
    </div>
);

//! STEPS TO CONNECT A COMPONENT TO THE REDUX STORE
// a) import { connect } from 'react-redux'
// 1. create a const for the higher order component by calling connect()
// 2. connect() returns a function, not a higher order component
// 3. to get the HOC, we need to call the function that connect returns, passing        in our component ->
// 4. connect()(ourComponent)
// 5. pass a function into connect that says what information from the store we         want our component to be able to access connect(()=> {})(ourComponent)
// 6. the argument to the anonymous function above is state. the information            returned is an object of key/value pairs usually from state, but we can           also add our own -> connect((state) => {name: 'harry'})(ourComponent)

//* export default connect()() is more common practice than const blahblah = connect()()
export default connect(state => {
    // what information from the store do we want our component to be able to access. the information will be returned as an object of key/value pairs

    return {
        // add a new property, like name
        name: 'cirrus',

        // or use the ones from state, like expenses and filters
        //expenses: state.expenses,
        //filters: state.filters,

        //* refactored to use the return value (the filtered sorted array) of getVisibleExpenses function, which was a default export, so here we imported it as selectExpenses

        expenses: selectExpenses(state.expenses, state.filters),
    };
})(ExpenseList);

//export default ConnectedExpenseList;

//* it is common practice to extract the function passed to connect() to it's own variable, const mapStateToProps, as follows:
/* 
    -named this way because it maps the store state to component props

    const mapStateToProps = (state) => {
        return {
            name: 'cirrus',
            expenses: state.expenses,
        };
    }

    - then connect becomes connect(mapStateToProps)(ourComponent)
    ! in summary, call connect, define the things we want to get off of the store, define the component we want to create the connected version of
    * the end result is a new component with the props from the store
*/
