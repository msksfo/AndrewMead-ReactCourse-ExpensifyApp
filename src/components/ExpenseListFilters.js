import React from 'react';
import { connect } from 'react-redux';
import {
    setTextFilter,
    sortByAmount,
    sortByDate,
    setStartDate,
    setEndDate,
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

// create a basic presentational component
//* the goal is to get the old value from the store, because we need to make sure the input always matches up with the text value on the redux store, so if that value changes via a dispatch call,  we want to read that value and use it here
// 1. connect ExpenseListFilters to the store

export class ExpenseListFilters extends React.Component {
    // set up state to track the focus for the date picker
    // it will either be null or a string
    state = {
        calendarFocused: null,
    };

    //* this function is called by the react dates library.
    // it's called with an object, containing a start date and end date
    onDatesChange = ({ startDate, endDate }) => {
        // dispatch the correct actions to get the filters to change
        //this.props.dispatch(setStartDate(startDate));
        //this.props.dispatch(setEndDate(endDate));

        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = calendarFocused => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = e => {
        // we have access to dispatch (as a prop) inside our connected components
        // use dispatch to update the store as we type
        // pass in the action object
        //this.props.dispatch(setTextFilter(e.target.value))

        this.props.setTextFilter(e.target.value);
    };

    onSortChange = e => {
        return e.target.value === 'date'
            ? this.props.sortByDate()
            : this.props.sortByAmount();
    };

    render() {
        return (
            <div>
                {/* provide an onChange handler so real time changes are dispatched from the component to the store */}
                <input
                    type="text"
                    //! use value and onChange so the component will be a controlled component
                    value={this.props.filters.text}
                    onChange={this.onTextChange}

                    /* this was refactored to not be an inline function
                    onChange={e => {
                        // we have access to dispatch (as a prop) inside our connected components
                        // use dispatch to update the store as we type
                        // pass in the action object
                        this.props.dispatch(setTextFilter(e.target.value));
                    }}
                    */
                />

                <select
                    //! add value here so that the select will be a controlled component!!!
                    //* it works without it, but in that case the DOM would be the single source of truth, rather than react, which is not what we want.
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                {/* allow user to pick the  range of dates they want to see  expenses for. the date range picker from moment.js takes five props (mando), two of which depend on state, so we will make this a class based component */}
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        filters: state.filters,
    };
};

// since dispatch is set up here, I will remove it up above in the method definitions
const mapDispatchToProps = dispatch => {
    return {
        setTextFilter: text => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: startDate => dispatch(setStartDate(startDate)),
        setEndDate: endDate => dispatch(setEndDate(endDate)),
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ExpenseListFilters);
