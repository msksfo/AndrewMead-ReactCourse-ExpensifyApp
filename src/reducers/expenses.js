const expensesReducerDefaultState = [];

// pass in the default state, an empty array. and pass in the action
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // add the expense to the array of expenses
            // return state.concat(action.expense) -> with es6 array spread ->
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(value => value.id !== action.id);

        //* the above return statement destructured ->
        // return state.filter({ id } => id !== action.id)

        case 'EDIT_EXPENSE':
            return state.map(value => {
                return value.id === action.id
                    ? { ...value, ...action.updates }
                    : value;
            });

        case 'SET_EXPENSES':
            // we don't care what is in state, or what the previous expenses were
            // we just want to set the expenses array completely
            return action.expenses;

        default:
            return state;
    }
};

export default expensesReducer;
