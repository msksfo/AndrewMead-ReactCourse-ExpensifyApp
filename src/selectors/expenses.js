import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    // get a subset of all the expenses
    return expenses
        .filter(value => {
            const createdAtMoment = moment(value.createdAt);

            const startDateMatch = startDate
                ? startDate.isSameOrBefore(createdAtMoment, 'day')
                : true;

            const endDateMatch = endDate
                ? endDate.isSameOrAfter(createdAtMoment, 'day')
                : true;

            // filter for expenses.description has the text variable string inside it
            // hint: use .includes() and convert both strings to lower case
            const textMatch = value.description
                .toLowerCase()
                .includes(text.toLowerCase());

            // add the expense to the filtered array if ALL three of these are true
            return startDateMatch && endDateMatch && textMatch;
        })
        .sort((a, b) => {
            if (sortBy === 'date') {
                // the bigger number is the more recently created
                return a.createdAt > b.createdAt ? -1 : 1;
            } else if (sortBy === 'amount') {
                // put most expensive expenses first
                return a.amount > b.amount ? -1 : 1;
            }
        });
};

export default getVisibleExpenses;
