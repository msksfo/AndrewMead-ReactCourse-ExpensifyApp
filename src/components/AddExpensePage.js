import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// export this (unconnected) so we can test it
export class AddExpensePage extends React.Component {
    onSubmit = expense => {
        // dispatch the action to the redux store by calling connect
        //props.dispatch(addExpense(expense));

        //* the above code was refactored to this after adding in mapDispatchToProps, which we did for testing purposes
        this.props.startAddExpense(expense);

        // force a redirect to the dashboard page when user submits
        // pass in the path to the page you want to go to
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => {
    // return an object with props. these props call dispatch
    return {
        startAddExpense: expense => dispatch(startAddExpense(expense)),
    };
};

export default connect(
    undefined,
    mapDispatchToProps
)(AddExpensePage);

/* 
     this was the original code: export default connect()(AddExpensePage);

    -because addExpense is a function we imported in, it is difficult to test (unlike if it were passed in as a prop). but connect gives us a work around. 

    so we will refactor the export to
    1.  export default connect(undefined, mapDispatchToProps)(AddExpensePage)

    2. note: we also changed the code inside the arrow function passed to onSubmit from this:  //props.dispatch(addExpense(expense))
                to this:  //props.onSubmit(expense)

    3. we changed it over to a class based component, to avoid inline functions

    the functionality is the same! it was only changed so we could test with jest



*/
