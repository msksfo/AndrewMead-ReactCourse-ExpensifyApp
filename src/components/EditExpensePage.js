import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';

// import my action generators for removing/editing expenses
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = expense => {
        this.props.startEditExpense(this.props.expense.id, expense);

        // history.push() navigates the user programatically, after some code has run. Link is an anchor tag, so nothing happens unless it is clicked
        this.props.history.push('/');
    };

    onRemove = () => {
        this.props.startRemoveExpense(this.props.expense.id);

        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm
                    // pass the expense object down (it was returned from  mapStateToProps), so that when the create expense page comes up, the field will be populated with the data that previously existed, so the user can edit it
                    expense={this.props.expense}
                    // dispatch the action to edit the expense
                    // redirect to the dashboard page
                    onSubmit={this.onSubmit}
                />

                <button
                    onClick={this.onRemove}
                    // we have access to dispatch (as a prop) inside our connected components
                    // use dispatch to update the store as users interact with our app
                    // pass in the action object
                    // remove an expense by dispatching an action when the button is clicke
                >
                    Remove
                </button>
            </div>
        );
    }
}

//* 1. react router (the Link that led us here) renders the higher order component
//* 2. the hoc passes the props through and allows us to add some new ones

const mapStateToProps = (state, props) => {
    //* render one new prop
    // use the passed in props to search through the expenses array
    return {
        expense: state.expenses.find(
            value => value.id === props.match.params.id
        ),
    };

    //! now the actual expense we are trying to edit comes back to us, so up above we can render an instance of Expense form, passing in this specific expense as a prop
};

//? why is props being passed in? i dont see that it's being used...
const mapDispatchToProps = (dispatch, props) => {
    return {
        // take the data that comes in, -----> pass it down
        startEditExpense: (id, expense) =>
            dispatch(startEditExpense(id, expense)),
        startRemoveExpense: data => dispatch(startRemoveExpense(data)),
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditExpensePage);
