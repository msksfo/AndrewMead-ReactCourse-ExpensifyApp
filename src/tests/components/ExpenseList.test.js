import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

//* when testing react components, we want to test the unconnected version, because we want to be able to provide a set of dynamic props
//! so, our props will not come from the store. we will provide them directly

test('should render ExpenseList with expenses', () => {
    //* render the component with the dummy data by:
    // 1. make variable 'wrapper by calling shallow()
    // 2. into the shallow call, we pass in the component we want to render,         ExpenseList
    // 3. ExpenseList expects one prop (which we called 'expenses') to be             passed in
    // 4. set the expenses prop equal to the array of dummy data, expenses

    const wrapper = shallow(<ExpenseList expenses={expenses} />);

    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);

    expect(wrapper).toMatchSnapshot();
});
