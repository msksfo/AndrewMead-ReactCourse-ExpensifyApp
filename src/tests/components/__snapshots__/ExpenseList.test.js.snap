// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    no expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={195}
    createdAt={0}
    description="candy"
    id={1}
    key="1"
  />
  <Connect(ExpenseListItem)
    amount={109500}
    createdAt={-345600000}
    description="rent"
    id={2}
    key="2"
  />
  <Connect(ExpenseListItem)
    amount={4500}
    createdAt={345600000}
    description="credit card"
    id={3}
    key="3"
  />
</div>
`;
